### PR TITLE
Check for privatelink and set as OCSP url if it exists

### PR DESCRIPTION
**Pull request by SimbaGithub**
Use privatelink for OCSP URL when the specified account is using privatelink.

### CODE REVIEW

**Comment by sfc-gh-hchaturvedi** [comment]
@SimbaGithub we also need to ensure that the ocsp responder url AND the ocsp cache server url both are updated to support privatelink. The URL received by `ocspRequestCallback` [here](https://github.com/snowflakedb/snowflake-connector-nodejs/blob/0c7eea1c3056c54f1a1a7b18b57a683393b8cb5b/lib/agent/check.js#L205) should also be updated for privatelink

**Comment by SimbaGithub** [comment]
> @SimbaGithub we also need to ensure that the ocsp responder url AND the ocsp cache server url both are updated to support privatelink. The URL received by `ocspRequestCallback` [here](https://github.com/snowflakedb/snowflake-connector-nodejs/blob/0c7eea1c3056c54f1a1a7b18b57a683393b8cb5b/lib/agent/check.js#L205) should also be updated for privatelink

Done. Both `SF_OCSP_RESPONDER_URL` and `SF_OCSP_RESPONSE_CACHE_SERVER_URL` are updated with the privatelink

**Review by sfc-gh-hchaturvedi — CHANGES_REQUESTED** [review]
@SimbaGithub OCSP URL and OCSP Cache Server URL are not the same. For privatelink OCSP URL can be built from privatelink OCSP CACHE SERVER URL by extracting the hostname and appending it with `retry/{actual ocsp url}/{actual ocsp request}` for example : 
`ocsp.abc.privatelink.snowflakecomputing.com/retry/ocsp.digicert.com/<BASE 64 OCSP Request>`
Consider how it's done in the python connector - https://github.com/snowflakedb/snowflake-connector-python/blob/aea85c48e1dcb5286bb6cefa45b537cd33a6cb20/src/snowflake/connector/ocsp_snowflake.py#L280

**Comment by SimbaGithub** [comment]
@sfc-gh-hchaturvedi  Updated to set the OCSP responder url similar to how the python connector sets it. Am I understanding correctly that Python's `OCSP_RETRY_URL` is equivalent to NodeJS' `SF_OCSP_RESPONDER_URL`?

Note:
NodeJS' OCSP request contains an extra field `requestExtensions` that Python's OCSP request does not.

Here they are for comparison:
- NodeJS OCSP request
```
req = {
    tbsRequest: 
        version: 'v1',
        requestList: [ {
		        reqCert: certID
        } ],
        requestExtensions: [ {
		        extnID: rfc2560['id-pkix-ocsp-nonce'],
		        critical: false,
		        extnValue: rfc2560.Nonce.encode(crypto.randomBytes(16), 'der')
        } ]
  };
```  
- Python OCSP request
```
ocsp_request = OCSPRequest({
    'tbs_request': TBSRequest({
        'version': Version(0),
        'request_list': Requests([
	        Request({
		        'req_cert': cert_id,
	        })]),
	}),
})
```
I checked Python's base64 value for OCSP request and it's very similar to NodeJS' base64 value but NodeJS is longer due to the extra field.

Also, `parsedUrl.protocol` already has a ':' so the extra ':' is removed.

**Comment by sfc-gh-hchaturvedi** [comment]
> @sfc-gh-hchaturvedi Updated to set the OCSP responder url similar to how the python connector sets it. Am I understanding correctly that Python's `OCSP_RETRY_URL` is equivalent to NodeJS' `SF_OCSP_RESPONDER_URL`?
> 
> Note:
> NodeJS' OCSP request contains an extra field `requestExtensions` that Python's OCSP request does not.
> 
> Here they are for comparison:
> 
> * NodeJS OCSP request
> 
> ```
> req = {
>     tbsRequest: 
>         version: 'v1',
>         requestList: [ {
> 		        reqCert: certID
>         } ],
>         requestExtensions: [ {
> 		        extnID: rfc2560['id-pkix-ocsp-nonce'],
> 		        critical: false,
> 		        extnValue: rfc2560.Nonce.encode(crypto.randomBytes(16), 'der')
>         } ]
>   };
> ```
> 
> * Python OCSP request
> 
> ```
> ocsp_request = OCSPRequest({
>     'tbs_request': TBSRequest({
>         'version': Version(0),
>         'request_list': Requests([
> 	        Request({
> 		        'req_cert': cert_id,
> 	        })]),
> 	}),
> })
> ```
> 
> I checked Python's base64 value for OCSP request and it's very similar to NodeJS' base64 value but NodeJS is longer due to the extra field.
> 
> Also, `parsedUrl.protocol` already has a ':' so the extra ':' is removed.

Since the extension's criticality is false setting or unsetting it shouldn't really make a difference. Anyways the fix looks good to me.  However let's add a test - let's force privatelink and mock an ocsp URL and a base64 request and check the final environment variable being exported for both the cache and the final SF_OCSP_RESPONDER_URL.

**Comment by SimbaGithub** [comment]
> @SimbaGithub Since the extension's criticality is false setting or unsetting it shouldn't really make a difference. Anyways the fix looks good to me. However let's add a test - let's force privatelink and mock an ocsp URL and a base64 request and check the final environment variable being exported for both the cache and the final SF_OCSP_RESPONDER_URL.

New OCSP tests are added. I used one of the existing test accounts and added ".privatelink" to the account name so there's no need for new environment variables.

Now it also checks if the `SF_OCSP_RESPONSE_CACHE_SERVER_URL` value contains 'ocsp' because some of the tests from _testConnectionWithOCSP.js_ sets the value for that env variable and would fail.

The `Account without privatelink` test connects with a non-privatelink account and checks that the OCSP variables are unchanged. The `Account with privatelink` test connects with the privatelink account then calls the check function with the mock url and mock ocsp request.

File: lib/connection/connection.js [hunk 1]
```diff
@@ -127,6 +139,11 @@ function Connection(context)
       !Util.exists(callback) || Util.isFunction(callback),
       ErrorCodes.ERR_CONN_CONNECT_INVALID_CALLBACK);
 
+    if (connectionConfig.host.endsWith(".privatelink.snowflakecomputing.com"))
```
Comment (sfc-gh-kwagner) [inline lib/connection/connection.js:142]: nit: make this into a constant and use the constant here and below

**Review by sfc-gh-kwagner — APPROVED** [review]
One minor comment but approved overall

File: lib/connection/connection.js (hunk 1 above)
Comment (SimbaGithub) [reply lib/connection/connection.js]: Ok, constant for privatelink suffix is now made

**Comment by SimbaGithub** [comment]
@sfc-gh-hchaturvedi just following up, please review when available thanks

**Review by sfc-gh-hchaturvedi — APPROVED** [review]
LGTM! Thanks @SimbaGithub .

**Comment by sfc-gh-hchaturvedi** [comment]
Please confirm test failure reasons before merging.

**Pull request merged**